File: frontend/src/resampler.ts
export interface ResampleInput {
  distance_m: number;
  hr_bpm: number | null;
  power_w: number | null;
  speed_mps: number | null;
  altitude_m: number | null;
}

export interface ResampleOutput {
  distance_m: number;
  hr_bpm: number | null;
  power_w: number | null;
  speed_mps: number | null;
  altitude_m: number | null;
}

export const BUCKET_SIZE_M = 50;

export function resampleByDistance(
  records: ResampleInput[],
  bucketSizeM: number = BUCKET_SIZE_M
): ResampleOutput[] {
  if (records.length === 0) return [];
  const maxDist = records[records.length - 1].distance_m;
  if (maxDist <= 0) return [records[0]];

  const numBuckets = Math.floor(maxDist / bucketSizeM);
  const buckets: ResampleOutput[] = [];

  for (let b = 0; b <= numBuckets; b++) {
    const targetDist = b * bucketSizeM;
    buckets.push(interpolateAt(records, targetDist));
  }

  return buckets;
}

function interpolateAt(
  records: ResampleInput[],
  targetDist: number
): ResampleOutput {
  if (records.length === 0) {
    return { distance_m: targetDist, hr_bpm: null, power_w: null, speed_mps: null, altitude_m: null };
  }

  if (targetDist <= records[0].distance_m) return snapTo(records[0], targetDist);

  const last = records[records.length - 1];
  if (targetDist >= last.distance_m) return snapTo(last, targetDist);

  let lo = 0;
  let hi = records.length - 1;
  while (lo < hi - 1) {
    const mid = Math.floor((lo + hi) / 2);
    if (records[mid].distance_m <= targetDist) lo = mid;
    else hi = mid;
  }

  const r0 = records[lo];
  const r1 = records[hi];
  const span = r1.distance_m - r0.distance_m;
  const t = span > 0 ? (targetDist - r0.distance_m) / span : 0;

  return {
    distance_m: targetDist,
    hr_bpm: lerp(r0.hr_bpm, r1.hr_bpm, t),
    power_w: lerp(r0.power_w, r1.power_w, t),
    speed_mps: lerp(r0.speed_mps, r1.speed_mps, t),
    altitude_m: lerp(r0.altitude_m, r1.altitude_m, t),
  };
}

function snapTo(r: ResampleInput, dist: number): ResampleOutput {
  return {
    distance_m: dist,
    hr_bpm: r.hr_bpm,
    power_w: r.power_w,
    speed_mps: r.speed_mps,
    altitude_m: r.altitude_m,
  };
}

function lerp(a: number | null, b: number | null, t: number): number | null {
  if (a === null && b === null) return null;
  if (a === null) return b;
  if (b === null) return a;
  return a + (b - a) * t;
}

File: backend/src/fitter/resampler.py
BUCKET_SIZE_M = 50


def resample_by_distance(
    records: list[dict],
    bucket_size_m: int = BUCKET_SIZE_M,
) -> list[dict]:
    if not records:
        return []
    max_dist = records[-1]["distance_m"]
    if max_dist <= 0:
        return [dict(records[0])]

    num_buckets = int(max_dist // bucket_size_m)
    buckets = []
    for b in range(num_buckets + 1):
        target = b * bucket_size_m
        buckets.append(_interpolate_at(records, target))
    return buckets


def _interpolate_at(records: list[dict], target_dist: float) -> dict:
    if not records:
        return {"distance_m": target_dist, "timestamp_s": 0.0}

    if target_dist <= records[0]["distance_m"]:
        return _with_distance(records[0], target_dist)

    last = records[-1]
    if target_dist >= last["distance_m"]:
        return _with_distance(last, target_dist)

    lo, hi = 0, len(records) - 1
    while lo < hi - 1:
        mid = (lo + hi) // 2
        if records[mid]["distance_m"] <= target_dist:
            lo = mid
        else:
            hi = mid

    r0, r1 = records[lo], records[hi]
    span = r1["distance_m"] - r0["distance_m"]
    t = (target_dist - r0["distance_m"]) / span if span > 0 else 0
    ts = _lerp(r0["timestamp_s"], r1["timestamp_s"], t)
    return {"distance_m": target_dist, "timestamp_s": ts}


def _with_distance(r: dict, dist: float) -> dict:
    return {"distance_m": dist, "timestamp_s": r["timestamp_s"]}


def _lerp(a: float, b: float, t: float) -> float:
    if a is None or b is None:
        return a if a is not None else b
    return a + (b - a) * t


def compute_time_gap_series(
    records_a: list[dict],
    records_b: list[dict],
    bucket_size_m: int = BUCKET_SIZE_M,
) -> list[dict]:
    resampled_a = resample_by_distance(records_a, bucket_size_m)
    resampled_b = resample_by_distance(records_b, bucket_size_m)

    min_len = min(len(resampled_a), len(resampled_b))
    series = []
    for i in range(min_len):
        gap = resampled_a[i]["timestamp_s"] - resampled_b[i]["timestamp_s"]
        series.append({"distance_m": resampled_a[i]["distance_m"], "gap_s": gap})
    return series